framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,901 @@
+        return scene.classes.vivify(className);
+    }
+
+    /**
+     * Returns true if am should not be inserted in source code, for example 
+     * {@link org.checkerframework.common.value.qual.BottomVal}. This happens
+     * when am cannot be inserted in source code or is the default for the
+     * location passed as argument.
+     * <p>
+     * Invisible qualifiers, which are annotations that contain the
+     * {@link org.checkerframework.framework.qual.InvisibleQualifier}
+     * meta-annotation, also return true.
+     * <p>
+     * TODO: Merge functionality somewhere else with
+     * {@link org.checkerframework.framework.type.GenericAnnotatedTypeFactory#createQualifierDefaults}.
+     * Look into the createQualifierDefaults method before changing anything here.
+     */
+    private static boolean shouldIgnore(AnnotationMirror am,
+            TypeUseLocation location, AnnotatedTypeFactory atf,
+            AnnotatedTypeMirror atm) {
+        AnnotationMirror bottomAnno = atf.getQualifierHierarchy().getBottomAnnotation(am);
+        if (AnnotationUtils.annotationName(bottomAnno) == AnnotationUtils.annotationName(am)) {
+            // Ignore annotation if it is the bottom type.
+            return true;
+        }
+        Element elt = am.getAnnotationType().asElement();
+        // Checks if am is an implementation detail (a type qualifier used
+        // internally by the type system and not meant to be seen by the user.)
+        Target target = elt.getAnnotation(Target.class);
+        if (target != null && target.value().length == 0) return true;
+        if (elt.getAnnotation(InvisibleQualifier.class) != null) return true;
+
+        // Checks if am is default
+        if (elt.getAnnotation(DefaultQualifierInHierarchy.class) != null) {
+            return true;
+        }
+        DefaultQualifier defaultQual = elt.getAnnotation(DefaultQualifier.class);
+        if (defaultQual != null) {
+            for (TypeUseLocation loc : defaultQual.locations()) {
+                if (loc == TypeUseLocation.ALL || loc == location) {
+                    return true;
+                }
+            }
+        }
+        DefaultFor defaultQualForLocation = elt.getAnnotation(DefaultFor.class);
+        if (defaultQualForLocation != null) {
+            for (TypeUseLocation loc : defaultQualForLocation.value()) {
+                if (loc == TypeUseLocation.ALL || loc == location) {
+                    return true;
+                }
+            }
+        }
+
+        // Checks if am is an implicit annotation
+        // TODO: Handles cases of implicit annotations added via an
+        // org.checkerframework.framework.type.treeannotator.ImplicitsTreeAnnotator
+        ImplicitFor implicitFor = elt.getAnnotation(ImplicitFor.class);
+        if (implicitFor != null) {
+            TypeKind[] types = implicitFor.types();
+            for (TypeKind tk : types) if (tk == atm.getKind()) return true;
```

Comment:
I would split this and other similar code over multiple lines and add more braces.